### PR TITLE
Compiler args performance enhancements

### PR DESCRIPTION
**Pull request by QuLogic**
See #1485 for test case; I assume there are no performance test cases at the moment. Without running the profiler, speeds are little bit better than I said before (averaged over 10 runs):
* master 80f870c: 18.05±0.05s
* ~~Use an `OrderedDict` when deduping d07b22f: 3.52±0.05s (×5.12±0.07)~~
* Use a `list`+`set` 128e2a6: 3.52±0.07s (×5.12±0.10)
* With a cache for basic compiler args ~~6d9d268~~ 11f8d4a: 2.52±0.04s (×7.16±0.13)

Profile now looks like this:
1000call-perf

You can make a _very_ small improvement (2.52±0.03s (×7.17±0.08)) by replacing `arg.startswith(prefixes)` with `arg[:2] in prefixes` in the `_can_dedup`/`_should_prepend` functions, but this assumes that all prefixes are 2 characters.

The caching is only based on global/subproject information; if the cache were made from all target properties, I'm sure things could be even faster, but I didn't feel like puzzling out what of the target to use in the key (since obviously using the target itself would be useless.)

### CODE REVIEW

File: mesonbuild/backend/backends.py [hunk 1]
```diff
@@ -333,15 +334,13 @@ def escape_extra_args(compiler, args):
...
-    def generate_basic_compiler_args(self, target, compiler, no_warn_args=False):
+    def _generate_basic_compiler_args(self, subproject, compiler, no_warn_args):
         # Create an empty commands list, and start adding arguments from
         # various sources in the order in which they must override each other
         # starting from hard-coded defaults followed by build options and so on.
         commands = CompilerArgs(compiler)
         # First, the trivial ones that are impossible to override.
         #
-        # Add -nostdinc/-nostdinc++ if needed; can't be overriden
-        commands += self.get_cross_stdlib_args(target, compiler)
```
Comment (QuLogic) [inline mesonbuild/backend/backends.py]: This appears to apply globally and so should be safe to move, I hope.

**Comment by QuLogic** [comment]
Err, why does the test case require that duplicate args be duplicated?

**Comment by QuLogic** [comment]
Sigh, I just realized the dict is over-dedup'ing. I might have to go with a `set`+`list` depending on how it times out.

**Comment by QuLogic** [comment]
`list`+`set` seems pretty close: 2.63±0.13s

**Comment by QuLogic** [comment]
@codido

**Comment by codido** [comment]
@QuLogic Thanks!

With the original example I'm seeing a huge boost, though as you mentioned, it still takes a couple of seconds. Not sure what would be a reasonable/feasible time, but running the same example with a single C file takes less than half a second, so guess things aren't completely cached just yet?

Also, for multiple include directories, such as the following example, I'm afraid I'm not seeing much difference with this (actually even runs a bit slower):
```
#!/bin/bash

cat > meson.build <<EOF
project('example', 'c')
inc = []
src = []
EOF

for i in {1..1000}
do
   mkdir -p dir$i
   touch file$i.c
   echo "inc += include_directories('dir$i')" >> meson.build
   echo "src += files('file$i.c')" >> meson.build
done

echo "executable('example', src, include_directories: inc)" >> meson.build

time meson build
rm -rf file*.c dir* build
```

**Comment by QuLogic** [comment]
Yes, it only caches based on compiler and subproject, but now that I think about it, maybe it can cache based on target as well if it needs to be applied to all the sources of the target.

**Comment by QuLogic** [comment]
Ah, I see it's triggered by extraneous stuff at a higher level; adding a cache at that point also helps: ~2.3s. I don't have time to totally test it now; maybe that cache is enough and the other changes aren't necessary...

Most of the time now is in output, which is full of string replacement for escaping. Maybe with `File` objects everywhere, this could be optimized away, but that's a large project.

**Comment by codido** [comment]
@QuLogic Looks much better now, thanks!

**Comment by nirbheek** [comment]
In addition to creating all these caches, I wonder if it makes sense to create per-target rules in the `build.ninja` file. Those are for exactly this purpose, and it would also:

1. Speed up writing of the Ninja file
2. Reduce the size of the `build.ninja` file enormously
3. Improve Ninja's parsing of the `build.ninja` file

We could in fact make things even nicer by having a rule like:

```ninja
incpaths = -Ifoo -Ibar
libpaths = -Lfdir -Lbdir
args = -e -v -e -r -y -t -h -i -n -g -else
rule sometarget
  command = gcc $incpaths $libpaths $in $args -o $out 
```

or similar.

(moved from the issue to here)

**Comment by QuLogic** [comment]
I had a chance to try out just the last commit, and since it's a top-level cache, both this case and the other issue are about the same, around 2.2 seconds. It's about the same or slightly slower (a few hundredths of a second) when the first two commits are also included. I didn't look at a spread of multiple runs to get any standard deviation or anything. Odd that it seems a little faster than last time, but I think this just getting into external noise at this point.

So I'm not sure whether we need to hang on to the first two commits, or whether they might be of some benefit for other pathological cases.

**Comment by jpakkane** [comment]
Simpler is better. If we can't get measurable proof that the changes make things faster, let's just drop them.

Per-target rules might make sense, but they complicate creating the compilation database a bit as it is generated for rules. But let's do that in a different MR in any case.

**Comment by QuLogic** [comment]
I'll hang on to them in case we need them, but I've removed the others from this PR.

**Comment by jpakkane** [comment]
Looking great, thanks.

**Pull request merged**